### ISSUE TITLE
Examples and classpaths

### ISSUE THREAD

**Issue by salotz** [kasperdokter/Reo#14]
I am not primarily a java dev so I apologize for any obvious ignorance, but I am following the installation instructions.

I am on Ubuntu 18.04 with:

```
--> java -version
openjdk version "1.8.0_242"
OpenJDK Runtime Environment (build 1.8.0_242-8u242-b08-0ubuntu3~18.04-b08)
OpenJDK 64-Bit Server VM (build 25.242-b08, mixed mode)
```

First I download the jar files:

```
reo-1.0.jar
reo-runtime-java.jar
reo-runtime-java-lykos.jar
```

And add them to my CLASSPATH as in the `initreo.sh` script that comes with the installer:

```
#!/bin/bash

alias reo="java -jar /home/salotz/Downloads/reo-linux_mac/bin/reo-1.0.jar"

export CLASSPATH="$CLASSPATH:.:/home/salotz/Downloads/reo-linux_mac/bin/reo-runtime-java.jar:/home/salotz/Downloads/reo-linux_mac/bin/reo-runtime-java-lykos.jar"

```

```
--> source initreo.sh
```

Then I cloned the repo so I could get all the examples (following the intro in the docs):

```
git clone git@github.com:ReoLanguage/Reo.git
cd Reo/examples/intro
```

```
--> reo main.treo 
line 9:16 token recognition error at: '#Ja'
line 10:18 token recognition error at: '#Ja'
line 11:17 token recognition error at: '#Ja'
[ERROR] (main.treo:9.22) mismatched input '"Processes.Red"' expecting {'(', '[', '<'}
[ERROR] (main.treo:10.24) mismatched input '"Processes.Green"' expecting {'(', '[', '<'}
[ERROR] (main.treo:11.23) mismatched input '"Processes.Blue"' expecting {'(', '[', '<'}
```

I also tried another one which can at least compile the `treo` file:

```
--> cd Reo/examples/test
--> reo filter_test.treo
--> javac filter_test.java 
filter_test.java:102: error: package nl.cwi.reo.templates does not exist
			           !(nl.cwi.reo.templates.Relations.Odd( $1.peek() ))));
			                                 ^
filter_test.java:108: error: package nl.cwi.reo.templates does not exist
			           nl.cwi.reo.templates.Relations.Odd( $1.peek() )));
			                               ^
filter_test.java:147: error: package nl.cwi.reo.templates does not exist
						     !(nl.cwi.reo.templates.Relations.Odd( $1.peek() )))) break;
						                           ^
filter_test.java:149: error: package nl.cwi.reo.templates does not exist
						     nl.cwi.reo.templates.Relations.Odd( $1.peek() ))) break;
						                         ^
4 errors
```

**Comment by k4r4c** [comment]
Hi! Thanks for the details, very much appreciated!
I would like first to ask you where did you find mention about cloning the repository for having the examples? There are some examples that can be compiled and run under the folder "./slides/main" in your installation directory. More examples can be found on the master branch, but this is not a prerequisite for introduction.

Now that you have cloned the master branch, some examples are out of date (the syntax evolved and the current parser cannot support those anymore). The file "Reo/examples/intro/main.treo" is one of them. The errors you get are parsing errors. Examples can be found in the documentation https://reo.readthedocs.io/en/latest/tutorial.html#components-and-interfaces.

Regarding the second file, the example uses an external function Odd from the java file Relations.java. The structure x.y.z is internal to Java as a package and filesystem representation. The file "filter_test.treo" gives to the component filter a String representing the path to the function Odd. You can change the path in order to link it to your own java class. The error only shows up when the generated java program is compiled: at this moment, one must link the class to a valid java class in order for the class to be compiled. A dynamic java classpath is given during compilation by the command javac -cp "$CLASSPATH:pathtofolder" file.java. The object must also be executed with the classpath as argument, i.e. java -cp "$CLASSPATH:pathtofolder" file .

**Comment by salotz** [comment]
Thanks for your patience. So the examples that come in the `archive/reo-linux_mac` should work then?

> I would like first to ask you where did you find mention about cloning the repository for having the examples?

I didn't but the README said to look in the examples folder which is only in the git repo.

> this is not a prerequisite for introduction.

In the introduction there is mention of the `Process.java` class which is in `examples/intro` so I also made an assumption it was the same.

Regarding the second example, I used the `-cp` option to set the classpath to the other class but then I got another about yet another class. Lets drop that examples since you say there may be some stale ones on the `master` branch.

I'm just trying to get something that compiles and runs so I can tweak it to help me understand how reo works. I read parts of the papers but I want to see it run to fully get it.

So if there is a best example which is it? The README from the `installer.sh` just says to run `reo FILE.treo -cp ../..`

If I go to `reo-linux_mac/slides/main` then choose one at random and compile it (after `source reo-linux_mac/initreo.sh`) I get:

```
--> echo $CLASSPATH
:.:/home/salotz/Downloads/reo-linux_mac/bin/reo-runtime-java.jar:/home/salotz/Downloads/reo-linux_mac/bin/reo-runtime-java-lykos.jar

--> alias reo
alias reo='java -jar /home/salotz/Downloads/reo-linux_mac/bin/reo-1.0.jar'

--> ls
blue.treo   main10.treo  main2.treo  main5.treo  main8.treo  red.treo
green.treo  main11.treo  main3.treo  main6.treo  main9.treo  workers
main1.treo  main12.treo  main4.treo  main7.treo  rba

--> reo blue.treo 

--> ls
blue.java   main1.treo   main12.treo  main4.treo  main7.treo  rba
blue.treo   main10.treo  main2.treo   main5.treo  main8.treo  red.treo
green.treo  main11.treo  main3.treo   main6.treo  main9.treo  workers

```

I get the `blue.java` file compiled from reo successfully.

Assuming the `CLASSPATH` env value is correct above then I try to compile the java code:

```
--> javac blue.java 
blue.java:45: error: cannot find symbol
			Consumer.blue($1);
			^
  symbol:   variable Consumer
  location: class blue1
1 error
```

Now I can see that there is a `Consumer.java` in the `workers` folder:

```
--> ls workers
Consumer.java  Functions.java  Producer.java
--> javac workers/Consumer.java 

--> ls workers
Consumer.class  Consumer.java  Functions.java  Producer.java

--> javac blue.java -cp "$CLASSPATH:./workers/"

--> ls
'blue$1.class'             blue.java    main10.treo   main3.treo   main7.treo   red.treo
'blue$PortWindow2.class'   blue.treo    main11.treo   main4.treo   main8.treo   workers
'blue$blue1.class'         green.treo   main12.treo   main5.treo   main9.treo
 blue.class                main1.treo   main2.treo    main6.treo   rba

--> java blue 
Exception in thread "Thread-0" java.lang.NoClassDefFoundError: Consumer
	at blue$blue1.run(blue.java:45)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: Consumer
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
	... 2 more
Exception in thread "Thread-1" java.awt.AWTError: Assistive Technology not found: org.GNOME.Accessibility.AtkWrapper
	at java.awt.Toolkit.loadAssistiveTechnologies(Toolkit.java:807)
	at java.awt.Toolkit.getDefaultToolkit(Toolkit.java:886)
	at sun.java2d.SunGraphicsEnvironment.getUsableBounds(SunGraphicsEnvironment.java:261)
	at java.awt.GraphicsEnvironment.getCenterPoint(GraphicsEnvironment.java:430)
	at sun.awt.X11GraphicsEnvironment.getCenterPoint(X11GraphicsEnvironment.java:318)
	at nl.cwi.reo.runtime.Windows.producer(Windows.java:48)
	at blue$PortWindow2.run(blue.java:61)
	at java.lang.Thread.run(Thread.java:748)
```

I can slowly figure this out piece by piece but isn't there like a build system to do this for me? How do you guys compile them?

And then just to be clear the final examples will be to run the compiled `mainX` files right?

**Comment by k4r4c** [comment]
> In the introduction there is mention of the Process.java class which is in examples/intro so I also made an assumption it was the same.

Yes, I see. I will change that part then, and relate it to an example in the archive. Thanks!

You're almost there. Java wants the link for compilation and for execution. You also need to provide the classpath when you are executing. Try this: `java blue.java -cp "$CLASSPATH:./workers/"`

You could use a Java environment, like Eclipse, that provides ways to link classes easily (using packages). You can also compile in Eclipse from Reo to Java into a runtime directory, and execute your compiled program with some runtime libraries. This is how I personally do when developing/testing the compiler.

Overall, what could make the process a bit confusing is that there are three levels of compilation. First, it goes from a Reo file to a Java program (this step could be modified to output a Promela program, or Rust), then it compiles the Java program using external library dynamically linked (the javac command), and then executed with the dynamically linked object (produced during compilation). The first step is the only one that is Reo specific. If reo is run without argument, a list of possible parameters are displayed in the terminal. One of the parameters is the target language (e.g. Java). The other steps depend on the target language chosen. 

I agree however. Understanding the whole process shouldn't be necessary for trying the tool and seeing some examples. Might be an idea to set a flag for a default compilation from Reo to Java binary.

**Comment by salotz** [comment]
> Try this: java blue.java -cp "$CLASSPATH:./workers/"

I did that in the next code block I was just documenting the process that someone would go through when trying to run the examples.

> Overall, what could make the process a bit confusing is that there are three levels of compilation.

This I actually understand, although it does make it hard to follow the examples. Which is why I wanted to run them.

The confusing issue for me so far is basically which stuff to compile, setting the classpath, and which files to actually run. Typically, projects have a makefile equivalent to do all this (and specify the relationships). Considering, there is an extra non-Java compilation step it seems to be especially important. I'm guessing Eclipse takes care of it for you automagically or you have some shell scripts you use.

I will probably end up writing a makefile or something similar just to get the examples running here. I don't really want to use Eclipse or any other IDE unless I must.

**Comment by k4r4c** [comment]
> I did that in the next code block I was just documenting the process that someone would go through when trying to run the examples.

Did it run? If this is not directly clear from the readme, I would have to emphasize more on that point. Note that the classpath must be provided before the java class or object: `java -cp ... filename.java` or `java -cp .... filename`.
On my side, following the steps in the readme (including the two last lines where it mentions external java classes) gives me some executable code.
For example:
```
$ cd slides/main/
$ reo main1.treo -cp ../..
$ javac -cp "$CLASSPATH:./workers/" main2.java
$ java -cp "$CLASSPATH:./workers/" main2
```

Building a good environment for compilation/execution on the user side is something that could be very useful, indeed.
As a proposition, you could equally put all the java function in the same folder, add the path to the CLASSPATH, and write a `make` script such as:
```
#!/bin/bash
java -jar /home/salotz/Downloads/reo-linux_mac/bin/reo-1.0.jar $1.treo -cp /home/salotz/Downloads/reo-linux_mac
javac -cp "$CLASSPATH:/home/salotz/Downloads/reo-linux_mac/slides/main/workers" $1.java  
java -cp "$CLASSPATH:/home/salotz/Downloads/reo-linux_mac/slides/main/workers" $1  
```
Executing the command `./make main3` in the folder  `/home/salotz/Downloads/reo-linux_mac/slides/main` should give you the outcome of the compiled `main3.treo` file.

**Comment by salotz** [comment]
Ah okay I see. You don't need to compile some of the treo files at all. For instance in `main1.reo`  you are importing from the treo file and not the compiled java files.

I was under the impression that all treo files need to be compiled. And since `main1.treo` is importing `blue.treo green.treo red.treo alternator.treo` they needed to be compiled individually.

------------------------

Also I meant makefiles as GNU make https://en.wikipedia.org/wiki/Make_(software).

You could just have a single makefile that compiles all the examples and then you can run each one.

I think this is better said with an example. Give me a bit now that I understand the whole process.

**Comment by salotz** [comment]
For example if you put the following makefile in the `slides/main` folder:
```
all: main1 main2
.PHONY: all

main1: main1.class
.PHONY: main1

main1.java:
	java -jar ../../bin/reo-1.0.jar main1.treo -cp ../..

main1.class: main1.java
	javac -cp "../../bin/reo-runtime-java.jar:../../bin/reo-runtime-java/jar:../../bin/reo-runtime-java-lykos.jar:./workers/:./workers/:../.." main1.java

main2: main2.class
.PHONY: main2

main2.java:
	java -jar ../../bin/reo-1.0.jar main2.treo -cp ../..

main2.class: main2.java
	javac -cp "../../bin/reo-runtime-java.jar:../../bin/reo-runtime-java/jar:../../bin/reo-runtime-java-lykos.jar:./workers/:./workers/:../.." main2.java


clean:
	rm -f *.java
	rm -f *.class
	rm -f *~
```

Then the instructions for one example (`main1`) would be:

```
source initreo.sh
alias run_reo='java -cp "$CLASSPATH:./workers/"'
cd slides/main
make main1
run_reo main1
```

And you could compile them all at once with: `make all` or `make` then `run_reo main2` etc.

If you edit the `treo` files then rerun `make` it will detect that and rerun things. You can also run `make clean` to get rid of all the compiled files.

I'm not a make expert but there are some features to make writing them for similar tasks like this easier but you get the idea.

**Comment by k4r4c** [comment]
Thanks for sharing the idea! Yes, Makefiles are amazing. This is valuable for a clean and easy deployment. I will include the feature in the next update.
Should I consider the thread closed?
